google/cloud/opentelemetry/internal/recordable.cc

```diff
@@ -13,15 +13,108 @@
 
 namespace google {
 namespace cloud {
 namespace otel_internal {
 GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
 namespace {
+
 // Taken from:
 // http://35.193.25.4/TensorFlow/models/research/syntaxnet/util/utf8/unilib_utf8_utils.h
 bool IsTrailByte(char x) { return static_cast<signed char>(x) < -0x40; }
+
+class AttributeVisitor {
+ public:
+  AttributeVisitor(
+      google::devtools::cloudtrace::v2::Span::Attributes& attributes,
+      opentelemetry::nostd::string_view key, std::size_t limit)
+      : attributes_(attributes), key_(key), limit_(limit) {}
+
+  void operator()(bool value) {
+    auto* proto = ProtoOrDrop();
+    if (!proto) return Drop();
+    proto->set_bool_value(value);
+  }
+  void operator()(std::int32_t value) {
+    auto* proto = ProtoOrDrop();
+    if (!proto) return Drop();
+    proto->set_int_value(value);
+  }
+  void operator()(std::uint32_t value) {
+    auto* proto = ProtoOrDrop();
+    if (!proto) return Drop();
+    proto->set_int_value(value);
+  }
+  void operator()(std::int64_t value) {
+    auto* proto = ProtoOrDrop();
+    if (!proto) return Drop();
+    proto->set_int_value(value);
+  }
+  void operator()(std::uint64_t value) {
+    auto* proto = ProtoOrDrop();
+    if (!proto) return Drop();
+    proto->set_int_value(value);
+  }
+  // The Cloud Trace proto does not accept floating point values, so we convert
+  // them to strings.
+  void operator()(double value) {
+    auto* proto = ProtoOrDrop();
+    if (!proto) return Drop();
+    SetTruncatableString(*proto->mutable_string_value(), absl::StrCat(value),
+                         kAttributeValueStringLimit);
+  }
+  void operator()(char const* value) {
+    auto* proto = ProtoOrDrop();
+    if (!proto) return Drop();
+    SetTruncatableString(*proto->mutable_string_value(), value,
+                         kAttributeValueStringLimit);
+  }
+  void operator()(opentelemetry::nostd::string_view value) {
+    auto* proto = ProtoOrDrop();
+    if (!proto) return Drop();
+    SetTruncatableString(*proto->mutable_string_value(), value,
+                         kAttributeValueStringLimit);
+  }
+  // There is no mapping from a `span<T>` to the Cloud Trace proto. We just drop
+  // these attributes.
+  template <typename T>
+  void operator()(opentelemetry::nostd::span<T>) {
+    Drop();
+  }
+
+ private:
+  // Returns nullptr if we drop the attribute. Otherwise, returns an
+  // AttributeValue proto to set.
+  google::devtools::cloudtrace::v2::AttributeValue* ProtoOrDrop() {
+    // We drop attributes whose keys are too long.
+    if (key_.size() > kAttributeKeyStringLimit) return nullptr;
+
+    // We do not do any sampling. We just accept the first N attributes we are
+    // given, and discard the rest. We may want to consider reservoir sampling
+    // in the future. See: https://en.wikipedia.org/wiki/Reservoir_sampling
+    auto& map = *attributes_.mutable_attribute_map();
+    if (attributes_.attribute_map().size() < limit_) {
```

Comment:
Nit: `if (map.size() < limit_)`.

Reply:
Lol... I need to think before I `git push`